### PR TITLE
batch channel opens

### PR DESCRIPTION
**Pull request by johncantrell97**
Adds a `ChannelOpener` struct that orchestrates batch channel opens.  A batch channel open is one where you can open multiple channels using a single on-chain transaction.

Right now LDK doesn't "officially" support this (https://github.com/lightningdevkit/rust-lightning/issues/1510) but was able to work around these issues with a few workarounds.

Added a new NodeRequest::OpenChannels that lets you provide a Vec of channels to open.  Also updated NodeRequest::OpenChannel to use batch channel open flow with a batch of size 1.

The [wip] part of this is around the actual API.  Right now there is no actual api exposed.  The other issue is around how to handle (and respond) when _some_ of the channels fail to open for whatever reason.

Right now the workflow keeps track of all open requests individually and the plan is to return list of tuples like this: Vec<(open_request,open_result)>

### CODE REVIEW

File: senseicore/src/chain/broadcaster.rs [hunk 1]
```diff
@@ -1,21 +1,45 @@
...
 use crate::events::SenseiEvent;
 
 use super::database::WalletDatabase;
-use bitcoin::Transaction;
+use bitcoin::{Transaction, Txid};
 use lightning::chain::chaininterface::BroadcasterInterface;
 use tokio::sync::broadcast;
 
 pub struct SenseiBroadcaster {
+    pub debounce: Mutex<HashMap<Txid, usize>>,
```
Comment (devrandom) [inline senseicore/src/chain/broadcaster.rs:14]: maybe `funding_batch`?

File: senseicore/src/chain/broadcaster.rs [hunk 2]
```diff
@@ -1,21 +1,45 @@
...
+        Self {
+            node_id,
+            broadcaster,
+            wallet_database,
+            event_sender,
+            debounce: Mutex::new(HashMap::new()),
+        }
+    }
+
+    pub fn set_debounce(&self, txid: Txid, count: usize) {
```
Comment (devrandom) [inline senseicore/src/chain/broadcaster.rs:37]: maybe `add_funding_batch`?

File: senseicore/src/channels.rs [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+                        {
+                            if *user_channel_id == request.custom_id {
+                                return true;
+                            }
+                        }
+                        false
+                    });
+
+                    if event.is_none() {
+                        return (request, Err(Error::FundingGenerationNeverHappened));
```
Comment (devrandom) [inline senseicore/src/channels.rs:145]: could the event happen later (after `wait_for_events`) if the peer is temporarily disconnected?

File: senseicore/src/channels.rs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                    }
+                }
+                (request, result)
+            })
+            .collect::<Vec<(ChannelOpenRequest, Result<[u8; 32], Error>)>>();
```
Comment (devrandom) [inline senseicore/src/channels.rs:150]: you might be able to replace with `collect::<Vec<_>>` to be less verbose

**Review by devrandom — COMMENTED** [review]
I know this is a draft, but I was looking anyway :)

File: senseicore/src/channels.rs (hunk 4 above)
Comment (johncantrell97) [reply senseicore/src/channels.rs:150]: yeah this worked, thanks

**Comment by johncantrell97** [comment]
Just updated this with a unified open channel api.  Instead of having two endpoints there will just be a single one that accepts a list of channel opens.  The client libraries can expose helpers for working with single vs batch channel opens.

Going to update the JS library and the web ui next.

**Pull request merged**